### PR TITLE
Allow to acquire tip by LocalStateQuery protocol client.

### PR DESCRIPTION
**Pull request by coot**
* Change mini-protocol description
* Redefine codec, add a test which ensures that the new version of the codec is compatible with the previous one.
* Updated consensus, including the server.
* Added new supported version in `ouroboros-consensus-cardano`
As a bonus:
* `instance Monoid a => Monoid (IOSim s a)`

### CODE REVIEW

File: typed-protocols-examples/src/Network/TypedProtocol/Codec.hs [hunk 1]
```diff
@@ -450,3 +453,45 @@ prop_codec_binary_compat
   -> Bool
 prop_codec_binary_compat runM codecA codecB stokEq msgA =
      runM $ prop_codec_binary_compatM codecA codecB stokEq msgA
+
+
+-- | Compatibility between two codecs of the same protocol.  Encode a message
+-- with one codec and decode it with the other one, then compare if the result
+-- is the same as initial message.
+--
+prop_codecs_compatM
```
Comment (dcoutts) [inline typed-protocols-examples/src/Network/TypedProtocol/Codec.hs:462]: Yeah, very nice.

File: ouroboros-network/protocol-tests/Ouroboros/Network/Protocol/LocalStateQuery/Test.hs [hunk 2]
```diff
@@ -321,26 +368,32 @@ prop_codec
...
   -> Bool
 prop_codec_cbor msg =
-  runST (prop_codec_cborM codec msg)
+  runST (prop_codec_cborM (codec True) msg)
+
+prop_codec_V7_compatible
+  :: AnyMessageAndAgencyV7
+  -> Bool
+prop_codec_V7_compatible (AnyMessageAndAgencyV7 msg) =
+    runST (prop_codecs_compatM (codec False) (codec True) msg)
```
Comment (dcoutts) [inline ouroboros-network/protocol-tests/Ouroboros/Network/Protocol/LocalStateQuery/Test.hs:399]: Yep nice.

Perhaps more code duplication in the `AnyMessageAndAgencyV7` QC instance than necessary, but ok.

File: ouroboros-consensus/src/Ouroboros/Consensus/MiniProtocol/LocalStateQuery/Server.hs [hunk 3]
```diff
@@ -30,11 +33,12 @@ localStateQueryServer cfg getPastLedger getImmutablePoint =
...
 
-    handleAcquire :: Point blk
+    handleAcquire :: Maybe (Point blk)
                   -> m (ServerStAcquiring blk (Point blk) (Query blk) m ())
-    handleAcquire pt = do
-        (mPastLedger, immutablePoint) <-
-          atomically $ (,) <$> getPastLedger pt <*> getImmutablePoint
+    handleAcquire mpt = do
+        (pt, mPastLedger, immutablePoint) <- atomically $ do
+          pt <- maybe getTipPoint pure mpt
```
Comment (dcoutts) [inline ouroboros-consensus/src/Ouroboros/Consensus/MiniProtocol/LocalStateQuery/Server.hs:40]: :+1:

File: ouroboros-consensus/src/Ouroboros/Consensus/Network/NodeToClient.hs [hunk 4]
```diff
@@ -232,6 +236,7 @@ clientCodecs ccfg version = Codecs {
 
     , cStateQueryCodec =
         codecLocalStateQuery
+          (networkVersion >= NodeToClientV_8)
```
Comment (dcoutts) [inline ouroboros-consensus/src/Ouroboros/Consensus/Network/NodeToClient.hs:239]: :+1:

File: ouroboros-network/src/Ouroboros/Network/Protocol/LocalStateQuery/Codec.hs [hunk 5]
```diff
@@ -64,11 +66,18 @@ codecLocalStateQuery encodePoint  decodePoint
...
         CBOR.encodeListLen 2
      <> CBOR.encodeWord 0
      <> encodePoint pt
 
+    encode (ClientAgency TokIdle) (MsgAcquire Nothing) | canAcquireTip =
+        CBOR.encodeListLen 1
+     <> CBOR.encodeWord 8
+
+    encode (ClientAgency TokIdle) (MsgAcquire Nothing) =
+      error "encodeFailure: this version does not support acquiring tip"
```
Comment (dcoutts) [inline ouroboros-network/src/Ouroboros/Network/Protocol/LocalStateQuery/Codec.hs:79]:
```suggestion
      error $ "encodeFailure: local state query: using acquire without a "
           ++ "Point must be conditional on negotiating v8 of the "
           ++ "node-to-client protocol"
```

**Review by dcoutts — APPROVED** [review]
Looks great. Very neat, step by step.

File: ouroboros-network/protocol-tests/Ouroboros/Network/Protocol/LocalStateQuery/Test.hs (hunk 2 above)
Comment (coot) [reply ouroboros-network/protocol-tests/Ouroboros/Network/Protocol/LocalStateQuery/Test.hs:399]: I should  write them them the other way around, its easier to extend an instance than shrink.

Comment (coot) [reply ouroboros-network/protocol-tests/Ouroboros/Network/Protocol/LocalStateQuery/Test.hs:377]: Fixed.

**Comment by coot** [comment]
I left some work in `test-consensus` (including an `undefined`). I hope @nfrisby can help to finish this.

**Comment by coot** [comment]
bors merge

File: ouroboros-network/src/Ouroboros/Network/Protocol/LocalStateQuery/Codec.hs [hunk 6]
```diff
@@ -110,7 +128,10 @@ codecLocalStateQuery encodePoint  decodePoint
       case (stok, len, key) of
         (ClientAgency TokIdle, 2, 0) -> do
           pt <- decodePoint
-          return (SomeMessage (MsgAcquire pt))
+          return (SomeMessage (MsgAcquire (Just pt)))
+
+        (ClientAgency TokIdle, 1, 8) | canAcquireTip -> do
```
Comment (nfrisby) [inline ouroboros-network/src/Ouroboros/Network/Protocol/LocalStateQuery/Codec.hs:133]: Hmm. The negotiated version `V_8` means the sender was promised that `TokIdle, 1, 8` is valid message. So this `canAcquireTip` guard seems like we're potentially breaking that promise (I have not yet checked when exactly `canAcquireTip` is set to `False`.)

It's relatively new for me to be worrying about these serialization level concerns. Am I oversimplifying?

File: ouroboros-consensus-test/test-consensus/Test/Consensus/MiniProtocol/LocalStateQuery/Server.hs [hunk 7]
```diff
@@ -140,16 +143,19 @@ checkOutcome k chain = conjoin . map (uncurry checkResult)
            (property False)
         | otherwise
         -> tabulate "Acquired" ["AcquireFailurePointTooOld"] $ property True
+    checkResult Nothing = \case
+      Right _result -> tabulate "Acquired" ["Success"] True
+      Left  failure -> counterexample ("acuire tip point resulted in " ++ show failure) False
```
Comment (nfrisby) [inline ouroboros-consensus-test/test-consensus/Test/Consensus/MiniProtocol/LocalStateQuery/Server.hs:148]: "acuire" -> "acquired"

File: ouroboros-consensus/src/Ouroboros/Consensus/MiniProtocol/LocalStateQuery/Server.hs [hunk 8]
```diff
@@ -1,5 +1,6 @@
 {-# LANGUAGE NamedFieldPuns      #-}
 {-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TupleSections       #-}
```
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/MiniProtocol/LocalStateQuery/Server.hs:3]: Let's just use `(,,,) pt`.

File: ouroboros-network/src/Ouroboros/Network/Protocol/LocalStateQuery/Codec.hs (hunk 6 above)
Comment (nfrisby) [reply ouroboros-network/src/Ouroboros/Network/Protocol/LocalStateQuery/Codec.hs:133]: Ah, nevermind. I see that the flag abstracts the negotiated version. The data flow makes sense 👍

Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/Protocol/LocalStateQuery/Codec.hs:133]: `canAcuireTip = version >= NodeToClientV_8`, where `version` is the negotiated version.  This means that these tokens are only valid for `V_8`.  There's even a [test](https://github.com/input-output-hk/ouroboros-network/pull/2875/commits/10ad3839e0895afed77b564b44f887174b4b7cf9#diff-ae122f3aa2edae6db73be6733912179ecb1d396e4bd973a38046766c6d874ab6R373) (its actually implemented in another [commit](https://github.com/input-output-hk/ouroboros-network/pull/2875/commits/77798f56b23816a6627752ff3e0710cb67ba4fb7) in this pr) which validates that the two codec versions are compatible.

File: ouroboros-consensus-test/test-consensus/Test/Consensus/MiniProtocol/LocalStateQuery/Server.hs (hunk 7 above)
Comment (coot) [reply ouroboros-consensus-test/test-consensus/Test/Consensus/MiniProtocol/LocalStateQuery/Server.hs:148]: I'll fix this in another pr.

File: ouroboros-consensus/src/Ouroboros/Consensus/MiniProtocol/LocalStateQuery/Server.hs (hunk 8 above)
Comment (coot) [reply ouroboros-consensus/src/Ouroboros/Consensus/MiniProtocol/LocalStateQuery/Server.hs:3]: I'll fix this in another pr.

**Pull request merged**